### PR TITLE
Prevent dynamic collection/index creation

### PR DESCRIPTION
**Pull request by AnthonySendra**
# Description

This PR prevents users to create a collection or index with other routes, like "document/create", "document/createOrReplace", ...

# Related issue

* #794 

# TODO
* [x] Add test in elasticsearch:create
* [x] Add test in elasticsearch:createOrReplace
* [x] Add test in elasticsearch:replace
* [x] Add test in elasticsearch:createCollection (test only index exists)
* [x] Add test in elasticsearch:import (check every lines)
* [x] Update package.json with the right version of kuzzle-common-object (https://github.com/kuzzleio/kuzzle-common-objects/pull/53)

### CODE REVIEW

File: lib/services/elasticsearch.js [hunk 1]
```diff
@@ -332,6 +333,10 @@ class ElasticSearch extends Service {
     assertNoRouting(esRequest);
     assertWellFormedRefresh(this.config.apiVersion, esRequest);
 
+    if (!this.kuzzle.indexCache.exists(esRequest.index, esRequest.type)) {
+      return Bluebird.reject(new PreconditionError(`Index '${esRequest.index}' and/or collection '${esRequest.type}' don't exist`))
```
Comment (kuzzle) [inline lib/services/elasticsearch.js]: MINOR Add a semicolon at the end of this statement. [rule](http://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3ASemicolon)

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: lib/services/elasticsearch.js [hunk 2]
```diff
@@ -598,6 +611,10 @@ class ElasticSearch extends Service {
       [esRequest.type]: {}
     };
 
+    if (!this.kuzzle.indexCache.exists(esRequest.index)) {
+      return Bluebird.reject(new PreconditionError(`Index '${esRequest.index}' doesn't exist`))
```
Comment (kuzzle) [inline lib/services/elasticsearch.js]: MINOR Add a semicolon at the end of this statement. [rule](http://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3ASemicolon)

**Comment by kuzzle** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Comment by benoitvidis** [comment]
Looks this PR is not ready yet and was put to review by mistake.
Putting it back to wip

File: features/step_definitions/collections.js [hunk 3]
```diff
@@ -79,6 +79,25 @@ var apiSteps = function () {
   this.When(/I check if collection "(.*?)" exists on index "(.*?)"$/, function (collection, index, cb) {
     return stepUtils.getReturn.call(this, 'collectionExists', index, collection, cb);
   });
+
+  this.When(/I create a collection named "([^"]*)" in index "([^"]*)"$/, function (collection, index, callback) {
```
Comment (ballinette) [inline features/step_definitions/collections.js]: seems this action is not used in test scenarios yet

File: lib/api/core/models/security/role.js [hunk 4]
```diff
@@ -94,30 +94,7 @@ class Role {
...
-        action: 'createCollection',
-        index: request.input.resource.index,
-        collection: request.input.resource.collection
-      }, request.context);
-
-      promises.push(this.isActionAllowed(newRequest, kuzzle));
-    }
-
-    if (typeof actionRights === 'boolean') {
+    if (_.isBoolean(actionRights)) {
```
Comment (ballinette) [inline lib/api/core/models/security/role.js]: is there a reason to prefer `_.isBoolean` as `typeof` ?

File: lib/services/elasticsearch.js [hunk 5]
```diff
@@ -473,6 +482,10 @@ class ElasticSearch extends Service {
     assertNoRouting(esRequest);
     assertWellFormedRefresh(this.config.apiVersion, esRequest);
 
+    if (!this.kuzzle.indexCache.exists(esRequest.index, esRequest.type)) {
```
Comment (ballinette) [inline lib/services/elasticsearch.js]: not necessary for replace, as we check the document existence just in the lines below.
If the document exists, the index/collection necessary exists too.

File: test/services/implementations/elasticsearch.test.js [hunk 6]
```diff
@@ -394,28 +422,41 @@ describe('Test: ElasticSearch service', () => {
...
       elasticsearch.client.exists.returns(Bluebird.resolve(true));
       elasticsearch.client.index.returns(Bluebird.reject(error));
+      elasticsearch.kuzzle.indexCache.exists.returns(true);
 
       request.input.resource._id = createdDocumentId;
 
       return should(elasticsearch.replace(request)).be.rejectedWith(error);
     });
 
-    it('should throw a NotFoundError Exception if document already exists', done => {
```
Comment (ballinette) [inline test/services/implementations/elasticsearch.test.js]: This test should not be deleted

File: lib/api/core/models/security/role.js (hunk 4 above)
Comment (AnthonySendra) [inline lib/api/core/models/security/role.js]: Nop, when the branch was created the line was `_.isBoolean`. I will change it.

File: lib/services/elasticsearch.js (hunk 5 above)
Comment (AnthonySendra) [inline lib/services/elasticsearch.js]: Yes but I thought it was a good idea to notify the user the problem can comes from a typo or something in the index/collection and not from the rest of the request.
I prefer if the error is more precise about what the problem is. What do you think?

And the check is fast

Comment (ballinette) [inline lib/services/elasticsearch.js]: why not, but in that case, to be consistent, the check should be done on `update` method too ;)

**Comment by AnthonySendra** [comment]
@scottinet we can't configure ES for the user. What is your idea for setting `action.auto_create_index` to false?

**Comment by ballinette** [comment]
I guess it should be done by default in ES config files

File: lib/services/elasticsearch.js (hunk 5 above)
Comment (AnthonySendra) [inline lib/services/elasticsearch.js]: Done

**Comment by scottinet** [comment]
@AnthonySendra > my bad, I thought these options could be passed to the ES client constructor.

**Pull request merged**